WebAssembly/wcslib-8.1/utils/wcsware.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.40.4.
.TH WCSWARE "1" "July 2023" "wcsware 8.1" "User Commands"
.SH NAME
wcsware \- Extract FITS WCS keywords for an image
.SH SYNOPSIS
.B wcsware
[\fI<option>\fR]... [\fI<fitsfile>\fR]
.SH DESCRIPTION
wcsware extracts the WCS keywords for an image from the specified FITS
file, constructs wcsprm structs for each coordinate representation found
and performs a variety of operations using them.  It features a "lint"
capability as one of its more useful operations.
.PP
By default, all known extensions to the FITS WCS standard are allowed,
including deprecated usage.  However, in "lint" mode, strict conformance
to the standard is enforced.
.PP
The FITS file may be specified according to the syntax understood by
cfitsio, for example "file.fits.gz+1" refers to the first extension (the
second HDU) of a gzip'd FITS file.  Use "\-" or omit the file name for
input from stdin.
.SH OPTIONS
.TP
\fB\-a\fR<alt>
Specify an alternate coordinate representation (ignored if
there is only one).  Can also be specified as a 0\-relative
index in the range 0 to 26, where alternates are sequenced
alphabetically following the primary representation.
.TP
\fB\-b\fR
Use wcsbth() for primary image headers, normally wcspih()
is used.  (Implies \fB\-i\fR.)
.TP
\fB\-c\fR
Apply wcspcx() to the wcsprm struct to decompose CDi_ja
into PCi_ja and CDELTia.
.TP
\fB\-cp\fR
As for \fB\-c\fR, and also unscramble axes.
.TP
\fB\-C\fR
As for \fB\-c\fR, decompose CDi_ja if present, otherwise recompose
PCi_ja and CDELTia.
.TP
\fB\-Cp\fR
As for \fB\-C\fR, and also unscramble axes.
.TP
\fB\-f\fR
Apply wcsfix() to the header.
.TP
\fB\-h\fR<hdu>
Move to HDU number (1\-relative) which is expected to
contain an image array.  Overrides cfitsio extended
filename syntax.  Also useful for input from stdin.
.TP
\fB\-i\fR
Allow image header WCS keywords in binary table headers.
.TP
\fB\-l\fR
Validate (lint) the WCS keyrecords in the specified FITS
header for conformance to the WCS standard.  (Implies \fB\-s\fR.)
.TP
\fB\-m\fR
Apply wcstrim() to the wcsprm struct.
.TP
\fB\-o\fR
Use wcshdo() to translate the wcsprm struct into a FITS
header and print it.
.TP
\fB\-p\fR
Print the struct(s) using wcsprt() (default operation).
.TP
\fB\-P\fR
Same as \fB\-p\fR but don't print a default struct.
.TP
\fB\-r\fR
Require strict adherence to the FITS WCS standard, though
allowing the deprecated AIPS\-convention keywords, CROTAn,
EPOCH, and VELREF, and also some other deprecated usage.
(Must follow \fB\-l\fR if relaxed linting is required.)
.TP
\fB\-s\fR
Require strict adherence to the FITS WCS standard,
disallowing all deprecated features.
.TP
\fB\-t\fR
Terse (with \fB\-l\fR), report rejected WCS keyrecords only.
.TP
\fB\-v\fR
Verbose (with \fB\-l\fR), report recognised WCS keyrecords as well.
.TP
\fB\-w\fR
Convert world coordinates, obtained from stdin, to pixel
coordinates using wcss2p().
.TP
\fB\-x\fR
Convert pixel coordinates, obtained from stdin, to world
coordinates using wcsp2s().
.TP
\fB\-z\fR
Print the size of the wcsprm struct, including allocated
memory.


CHANGELOG.md
## [5.0.46] - 2023-10-11

* corrected the display of _*RA*_/_*DEC*_ coordinates (for example when moving the mouse over the FITS image), which brings the _*beta*_ v5 into line with FITSWebQL v4 and ds9

## [5.0.45] - 2023-10-04

* an improved P-V diagram: an _*all-new*_ crosshair with a tooltip (_*hint*_: hover a mouse over the P-V diagram)

## [5.0.44] - 2023-09-14

* spectrum binning: replaced the _*Largest Triangle Three Buckets*_ downsampling with true binning (averaging adjacent values)

## [5.0.43] - 2023-09-13

* added a new spectrum binning functionality (see a menu setting _*Preferences*_ / _*spectrum binning*_)

## [5.0.42] - 2023-09-11

* fixed a long-standing "_*disappearing menu*_" issue in Firefox (worked around a Firefox CSS hover bug)

## [5.0.41] - 2023-09-05

* re-worked the client-side image legend WebGL display to get around a bug (a slightly shifted legend position) for some combinations of Safari / macOS versions
* removed _*Google Analytics*_

## [5.0.40] - 2023-08-28

* an improved P-V Diagram: changed the X-axis to display an angular distance (in _*arcseconds*_) instead of a P-V line pixel number

## [5.0.39] - 2023-08-21

* corrected a velocity axis direction in the P-V Diagram for datasets with a reversed frequency / velocity axis
* added a user-moveable horizontal velocity reference line at _*v=0*_ (mid-velocity if there is no zero) to the P-V Diagram

## [5.0.38] - 2023-08-08

* corrected the auto-detection of radio FITS files (the previous version would incorrectly detect them as optical) for the newly added NRO45M datasets, based on the FITS header telescope keyword

## [5.0.37] - 2023-07-31

* updated the client/server ZFP compression library to v1.0.0
* started loading the client-side JavaScript library _*d3.js*_ from the global Content Delivery Network (CDN)
* added support for new _*NRO45M*_ datasets
* recompiled WebAssembly with the latest _*emscripten*_ compiler
* on-going work on the NOBEYAMA FUGIN composite RGB support

## [5.0.36] - 2023-05-30

* updated the mongoose C networking library to v7.10
* re-worked internal messaging to take advantage of the new mongoose v7.10 lock-free message queue (_*improved performance*_)
* applied glib atomic reference counting to the user session C structure (_*improved reliability*_)

## [5.0.35] - 2023-05-24

* Introduced a hard limit of 500 Splatalogue spectral lines that can be displayed simultaneously in the website. This has been done in order to prevent an information overload and a website slowdown. The 500 lines that are displayed will be chosen as a random subset of the larger than 500 original number of lines. Please refine your search criteria to avoid a hard limit. If the hard limit has been reached information will be displayed in the web browser JavaScript console (press F12 to view it in most browsers).
* On-going work on NOBEYAMA FUGIN

## [5.0.34] - 2023-04-25

* overhauled the server logging functionality (moved from FORTRAN to C)
* ensured compatibility with the new gcc / gfortran v13
* updated _*http.c*_ so that it compiles cleanly with the latest glib _*g_string_free(..., FALSE)*_ function

## [5.0.33] - 2023-04-21

* speeded up initial web page loading by using _*async*_ / _*await*_ within JavaScript
* switched loading of the 3D surface script from the local server over to the global CDN (Content Delivery Network)
* speeded up loading of contour lines _*new Worker(...)*_ scripts
* on-going work on the NOBEYAMA FUGIN composite RGB support

## [5.0.32] - 2023-04-13

* client-side JavaScript: improved WebGL reliability

## [5.0.31] - 2023-04-06

* replaced the ```html_encode()``` C function with a more robust ```mg_url_encode()```
* added initial support for loading multiple (composite) FITS files from the NOBEYAMA FUGIN project across the cluster
* added support for multiple (composite) FITS files to the WebSocket user session C structure
* propagating WebSocket heartbeats across the cluster using HTTP messages 

## [5.0.30] - 2023-03-29

* added compressed FITS files support to the external URL file downloads (_*.fits.Z_, _*.fits.gz_, _*.fits.zip_, _*.fits.bz2_)
* font colour adjustment in the P-V Diagram _#contour levels_ input field
* shortened the _source code_ entry in the web browser welcome screen

## [5.0.29] - 2023-03-15

* re-worked the client-side JavaScript P-V Diagram event loop to reduce the load on the server

## [5.0.28] - 2023-03-15

* improved the colourmap in the P-V Diagram (removed the _erf()_ scaling step)
* pixel values are now simply standardised / capped to the range [-3,3] prior to scaling to [0,1] for the colourmap
* adjusted the font colour in the P-V Diagram _#contour levels_ input field

## [5.0.27] - 2023-03-10

* adjusted the font family used in the welcome screen _CHANGELOG_

## [5.0.26] - 2023-03-10

* added a _CHANGELOG_ to the web browser welcome screen

## [5.0.25] - 2023-03-08

Added the ability to change the number of contour level as well as a choice of three contour-level scaling functions to the P-V Diagram:
* _linear_ (default)
* _sqrt_
* _logarithmic_
    
## [5.0.24] - 2023-03-06

* added contouring (a _linear_ scale only at the moment) to the P-V Diagram with a corresponding checkbox to enable/disable contouring
* improved serving the WASM files via the CDN (added the release numbers for better version control)
    
## [5.0.23] - 2023-03-02


WebAssembly/wcslib-8.1/utils/wcsware.c
/*============================================================================
  WCSLIB 8.1 - an implementation of the FITS WCS standard.
  Copyright (C) 1995-2023, Mark Calabretta

  This file is part of WCSLIB.

  WCSLIB is free software: you can redistribute it and/or modify it under the
  terms of the GNU Lesser General Public License as published by the Free
  Software Foundation, either version 3 of the License, or (at your option)
  any later version.

  WCSLIB is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
  more details.

  You should have received a copy of the GNU Lesser General Public License
  along with WCSLIB.  If not, see http://www.gnu.org/licenses.

  Author: Mark Calabretta, Australia Telescope National Facility, CSIRO.
  http://www.atnf.csiro.au/people/Mark.Calabretta
  $Id: wcsware.c,v 8.1 2023/07/05 17:12:07 mcalabre Exp $
*=============================================================================
* Usage: wcsware [<option>]... [<fitsfile>]
*-----------------------------------------------------------------------------
* wcsware extracts the WCS keywords for an image from the specified FITS file,
* constructs wcsprm structs for each coordinate representation found and
* performs a variety of operations using them.  Refer to the usage notes
* below.
*---------------------------------------------------------------------------*/

char usage[] =
"Usage: wcsware [<option>]... [<fitsfile>]\n"
"\n"
"wcsware extracts the WCS keywords for an image from the specified FITS\n"
"file, constructs wcsprm structs for each coordinate representation found\n"
"and performs a variety of operations using them.  It features a \"lint\"\n"
"capability as one of its more useful operations.\n\n"
"By default, all known extensions to the FITS WCS standard are allowed,\n"
"including deprecated usage.  However, in \"lint\" mode, strict conformance\n"
"to the standard is enforced.\n\n"
"The FITS file may be specified according to the syntax understood by\n"
"cfitsio, for example \"file.fits.gz+1\" refers to the first extension (the\n"
"second HDU) of a gzip'd FITS file.  Use \"-\" or omit the file name for\n"
"input from stdin.\n"
"\n"
"Options:\n"
"  -a<alt>      Specify an alternate coordinate representation (ignored if\n"
"               there is only one).  Can also be specified as a 0-relative\n"
"               index in the range 0 to 26, where alternates are sequenced\n"
"               alphabetically following the primary representation.\n"
"  -b           Use wcsbth() for primary image headers, normally wcspih()\n"
"               is used.  (Implies -i.)\n"
"  -c           Apply wcspcx() to the wcsprm struct to decompose CDi_ja\n"
"               into PCi_ja and CDELTia.\n"
"  -cp          As for -c, and also unscramble axes.\n"
"  -C           As for -c, decompose CDi_ja if present, otherwise recompose\n"
"               PCi_ja and CDELTia.\n"
"  -Cp          As for -C, and also unscramble axes.\n"
"  -f           Apply wcsfix() to the header.\n"
"  -h<hdu>      Move to HDU number (1-relative) which is expected to\n"
"               contain an image array.  Overrides cfitsio extended\n"
"               filename syntax.  Also useful for input from stdin.\n"
"  -i           Allow image header WCS keywords in binary table headers.\n"
"  -l           Validate (lint) the WCS keyrecords in the specified FITS\n"
"               header for conformance to the WCS standard.  (Implies -s.)\n"
"  -m           Apply wcstrim() to the wcsprm struct.\n"
"  -o           Use wcshdo() to translate the wcsprm struct into a FITS\n"
"               header and print it.\n"
"  -p           Print the struct(s) using wcsprt() (default operation).\n"
"  -P           Same as -p but don't print a default struct.\n"
"  -r           Require strict adherence to the FITS WCS standard, though\n"
"               allowing the deprecated AIPS-convention keywords, CROTAn,\n"
"               EPOCH, and VELREF, and also some other deprecated usage.\n"
"               (Must follow -l if relaxed linting is required.)\n"
"  -s           Require strict adherence to the FITS WCS standard,\n"
"               disallowing all deprecated features.\n"
"  -t           Terse (with -l), report rejected WCS keyrecords only.\n"
"  -v           Verbose (with -l), report recognised WCS keyrecords as well.\n"
"  -w           Convert world coordinates, obtained from stdin, to pixel\n"
"               coordinates using wcss2p().\n"
"  -x           Convert pixel coordinates, obtained from stdin, to world\n"
"               coordinates using wcsp2s().\n"
"  -z           Print the size of the wcsprm struct, including allocated\n"
"               memory.\n";

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <fitsio.h>

#include <tab.h>
#include <wcs.h>
#include <wcshdr.h>
#include <wcsfix.h>
#include <wcsprintf.h>
#include <getwcstab.h>

int main(int argc, char **argv)

{
  // Parse options.
  int allimg =  0;
  int ctrl   = -3;
  int dobth  =  0;
  int dofix  =  0;
  int dohdr  =  0;
  int dolint =  0;
  int dopcx  =  0;
  int doperm =  0;
  int dopix  =  0;
  int doprt  =  0;
  int dosize =  0;
  int dotrim =  0;
  int dowrld =  0;
  int hdunum =  0;
  int relax  = WCSHDR_all;
  int strict =  0;

  char *alt = 0x0;

  int iopt;
  for (iopt = 1; iopt < argc && argv[iopt][0] == '-'; iopt++) {
    if (!argv[iopt][1]) break;

    switch (argv[iopt][1]) {
    case 'a':
      // Select an alternate WCS.
      alt = argv[iopt]+2;
      break;

    case 'b':
      // Use wcsbth() for image headers.
      dobth  = 1;
      allimg = 1;
      break;

    case 'c':
      // Apply wcspcx().
      dopcx  = 1;
      doperm = (argv[iopt][2] == 'p');
      break;

    case 'C':
      // Apply wcspcx() with dopc == 1.
      dopcx  = 2;
      doperm = (argv[iopt][2] == 'p');
      break;

    case 'f':
      // Apply wcsfix().
      dofix  = 1;
      break;

    case 'h':
      // Move to HDU number.
      hdunum = atoi(argv[iopt]+2);
      if (hdunum < 0) {
        hdunum = 0;
        wcsfprintf(stderr, "wcsware: Invalid HDU number ignored.\n");
      }
      break;

    case 'i':
      // Allow image header keywords in bintables.
      allimg = 1;
      break;

    case 'l':
      // WCS lint.
      strict = 1;
      relax  = WCSHDR_strict;
      ctrl   = 3;
      dolint = 1;
      break;

    case 'm':
      // Apply wcstrim().
      dotrim = 1;
      break;

    case 'o':
      // Print header using wcshdo().
      dohdr = 1;
      break;

    case 'p':
      // Print structs.
      doprt  = 1;
      break;

    case 'P':
      // Print non-default structs.
      doprt  = -1;
      break;

    case 'r':
      // Strict, but allow some deprecated usage.
      strict = 0;
      relax  = WCSHDR_reject;
      break;

    case 's':
      // Strict, really strict.
      strict = 1;
      relax  = WCSHDR_strict;
      break;

    case 't':
      // Terse.
      ctrl   = 2;
      break;

    case 'v':
      // Verbose.
      ctrl   = 4;
      break;

    case 'w':
      // Convert world coordinates to pixel coordinates.
      dowrld = 1;
      break;

    case 'x':
      // Convert pixel coordinates to world coordinates.
      dopix  = 1;
      break;

    case 'z':
      // Print the full size of the wcsprm struct.
      dosize = 1;
      break;

    default:
      wcsfprintf(stderr, "%s", usage);
      return 1;
    }
  }

  if (allimg) {
    relax |= WCSHDR_ALLIMG;
  }

  char *infile;
  if (iopt < argc) {
    infile = argv[iopt++];

    if (iopt < argc) {
      wcsfprintf(stderr, "%s", usage);
      return 1;
    }
  } else {
    infile = "-";
  }

  if (!(dolint || dohdr || doprt || dopix || dowrld)) {
    doprt  = 1;
    dotrim = 1;
  }


  // Open the FITS file and move to the required HDU.
  int fitstat = 0;
  fitsfile *fptr;
  if (fits_open_file(&fptr, infile, READONLY, &fitstat)) goto cleanup;

  int hdutype;
  if (hdunum) {
    if (fits_movabs_hdu(fptr, hdunum, &hdutype, &fitstat)) goto cleanup;
  } else {
    if (fits_get_hdu_type(fptr, &hdutype, &fitstat)) goto cleanup;
  }

  // Read in the FITS header, excluding COMMENT and HISTORY keyrecords.
  char *header;
  int nkeyrec;
  if (fits_hdr2str(fptr, 1, NULL, 0, &header, &nkeyrec, &fitstat)) {
    goto cleanup;
  }


  // Interpret the WCS keywords.
  int keysel, nreject, nwcs, status;
  struct wcsprm *wcs;
  if ((hdutype == BINARY_TBL) ||
     ((hdutype == IMAGE_HDU) && dobth)) {
    keysel = (hdutype == BINARY_TBL) ? 0 : WCSHDR_IMGHEAD;
    if ((status = wcsbth(header, nkeyrec, relax, ctrl, keysel, 0x0, &nreject,
                    &nwcs, &wcs))) {
      wcsfprintf(stderr, "wcsbth ERROR %d: %s.\n", status,
                 wcshdr_errmsg[status]);
      goto cleanup;
    }
  } else if (hdutype == IMAGE_HDU) {
    if ((status = wcspih(header, nkeyrec, relax, ctrl, &nreject,
                    &nwcs, &wcs))) {
      wcsfprintf(stderr, "wcspih ERROR %d: %s.\n", status,
                 wcshdr_errmsg[status]);
      goto cleanup;
    }
  } else {
    wcsfprintf(stderr, "wcsware: Invalid FITS extension type.\n");
    goto cleanup;
  }

  free(header);

  if (nreject) {
    if (ctrl <= 3) {
      wcsprintf("\n%d WCS keyrecords were rejected.\n", nreject);
    }
    wcsprintf("\nThe rejected keyrecords do not conform%s to the FITS WCS "
              "standard.\n", strict?" strictly":"");
  } else if (nwcs == 0) {
    if (2 < ctrl) wcsprintf("\n");
    wcsprintf("No world coordinate systems found.\n");
    fits_close_file(fptr, &fitstat);
    return 0;

  } else if (2 < ctrl) {
    wcsprintf("\nNo invalid WCS keyrecords were found.\n");
  }


  // Sort out alternates.
  int  alts[27];
  struct wcsprm *wcsi = 0x0;
  if (alt) {
    int i = 0;

    if ('0' <= *alt && *alt <= '9') {
      if ((i = atoi(alt)) > nwcs-1) {
        wcsfprintf(stderr, "WARNING, no alternate coordinate "
          "representation \"%s\".\n", alt);
        goto cleanup;
      }

    } else {
      wcsidx(nwcs, &wcs, alts);

      int ialt = toupper(*alt);
      if (strlen(alt) > 1) {
        wcsfprintf(stderr, "WARNING, alternate specifier \"%s\" is "
          "invalid.\n", alt);
        goto cleanup;

      } else if (*alt == ' ') {
        if (alts[0] == -1) {
          wcsfprintf(stderr, "WARNING, no primary coordinate "
            "representation.\n");
          goto cleanup;
        }

      } else if (ialt < 'A' || ialt > 'Z') {
        wcsfprintf(stderr, "WARNING, alternate specifier \"%s\" is "
          "invalid.\n", alt);
        goto cleanup;

      } else {
        if ((i = alts[ialt - 'A' + 1]) == -1) {
          wcsfprintf(stderr, "WARNING, no alternate coordinate "
            "representation \"%s\".\n", alt);
          goto cleanup;
        }
      }
    }

    wcsi = wcs + i;
  }


  // Initialize and possibly print the structs.
  wcserr_enable(1);
  wcsprintf_set(stdout);

  for (int i = 0; i < nwcs; i++) {
    if (wcsi) {
      if ((wcs+i) != wcsi) continue;
    } else if (i) {
      wcsprintf("\nType <CR> for next: ");
      fgetc(stdin);
      wcsprintf("\n");
    }

    // Read -TAB arrays from the binary table extension (if necessary).
    if (fits_read_wcstab(fptr, wcs[i].nwtb, (wtbarr *)wcs[i].wtb,
                         &fitstat)) {
      goto cleanup;
    }

    // Translate non-standard WCS keyvalues?
    if (dofix) {
      int *stat = malloc(NWCSFIX * sizeof(int));
      if ((status = wcsfix(7, 0, wcs+i, stat))) {
        for (int j = 0; j < NWCSFIX; j++) {
          if (stat[j] > 0) {
             wcsfprintf(stderr, "wcsfix ERROR %d: %s.\n", status,
                        wcsfix_errmsg[stat[j]]);
          }
        }

        goto cleanup;
      }

      free(stat);
    }

    if ((status = wcsset(wcs+i))) {
      wcsperr(wcs+i, "");
      continue;
    }

    // Decompose CDi_ja, or recompose PCi_ja and CDELTia?
    if (dopcx) {
      if (dopcx == 1 || !((wcs+i)->altlin & 1)) {
        wcsprintf("Decomposing CDi_ja into PCi_ja and CDELTia");
      } else {
        wcsprintf("Recomposing PCi_ja and CDELTia");
      }

      if (doperm) {
        wcsprintf(" with axis permutation");
      }
      wcsprintf(".\n");

      double rotn[2];
      if (wcspcx(wcs+i, dopcx-1, doperm, rotn)) {
        wcsperr(wcs+i, "");
        continue;
      }

      wcsprintf("Rotation of celestial axis basis vectors (deg): "
        "%10.5f, %10.5f\n", rotn[0], rotn[1]);
    }

    // Apply wcstrim()?
    if (dotrim) {
      if (wcstrim(wcs+i)) {
        wcsperr(wcs+i, "");
        continue;
      }
    }

    // Extract WCSNAME from the wcsprm struct.
    char wcsname[72];
    strcpy(wcsname, wcs[i].wcsname);

    // Print the struct as a FITS header?
    if (dohdr) {
      if (wcshdo(WCSHDO_all, wcs+i, &nkeyrec, &header)) {
        wcsperr(wcs+i, "");
        continue;
      }

      char *hptr = header;
      wcsprintf("\n");
      for (int k = 0; k < nkeyrec; k++, hptr += 80) {
        wcsprintf("%.80s\n", hptr);
      }

      free(header);
    }

    // Print the struct?
    if (doprt) {
      if (doprt == 1 || strcmp(wcsname, "DEFAULTS")) {
        if (strlen(wcsname)) {
          wcsprintf("\n%s\n", wcsname);
        }

        wcsprintf("\n");
        wcsprt(wcs+i);
      }
    }

    // Print full size of the struct.
    if (dosize) {
      int sizes[2];
      wcssize(wcs+i, sizes);
      wcsprintf("\nSize of wcsprm struct: %4d (base), plus allocated memory: "
        "%4d, total: %4d (bytes).\n", sizes[0], sizes[1], sizes[0] + sizes[1]);
      wcsprintf("Constituent structs (prjprm is included in celprm):\n");

      auxsize((wcs+i)->aux, sizes);
      wcsprintf("        auxprm struct: %4d\n", sizes[0]);

      for (int itab = 0; itab < (wcs+i)->ntab; itab++) {
        tabsize((wcs+i)->tab + itab, sizes);
        wcsprintf("        tabprm struct: %4d (base), plus allocated memory: "
          "%4d, total: %4d (bytes).\n", sizes[0], sizes[1],
          sizes[0] + sizes[1]);
      }

      linsize(&((wcs+i)->lin), sizes);
      wcsprintf("        linprm struct: %4d (base), plus allocated memory: "
        "%4d, total: %4d (bytes).\n", sizes[0], sizes[1],
        sizes[0] + sizes[1]);

      celsize(&((wcs+i)->cel), sizes);
      wcsprintf("        celprm struct: %4d (base), plus allocated memory: "
        "%4d, total: %4d (bytes).\n", sizes[0], sizes[1],
        sizes[0] + sizes[1]);

      prjsize(&((wcs+i)->cel.prj), sizes);
      wcsprintf("        prjprm struct: %4d (base), plus allocated memory: "
        "%4d, total: %4d (bytes).\n", sizes[0], sizes[1],
        sizes[0] + sizes[1]);

      spcsize(&((wcs+i)->spc), sizes);
      wcsprintf("        spcprm struct: %4d (base), plus allocated memory: "
        "%4d, total: %4d (bytes).\n", sizes[0], sizes[1],
        sizes[0] + sizes[1]);
    }


    // Transform coordinates?
    if (dopix || dowrld) {
      if (strlen(wcsname)) {
        wcsprintf("\n%s\n", wcsname);
      }

      double *world  = 0x0;
      double *imgcrd = 0x0;
      double *pixcrd = 0x0;
      int    *stat   = 0x0;

      int nelem = wcs[i].naxis;
      world  = realloc(world,  nelem * sizeof(double));
      imgcrd = realloc(imgcrd, nelem * sizeof(double));
      pixcrd = realloc(pixcrd, nelem * sizeof(double));
      stat   = realloc(stat,   nelem * sizeof(int));

      if (dopix) {
        // Transform pixel coordinates.
        while (1) {
          wcsprintf("\nEnter %d pixel coordinate element%s: ", nelem,
            (nelem==1)?"":"s");
          int c = fgetc(stdin);
          if (c == EOF || c == '\n') {
            if (c == EOF) wcsprintf("\n");
            break;
          }
          ungetc(c, stdin);

          int ok;
          if ((ok = (scanf("%lf", pixcrd) == 1))) {
            for (int j = 1; j < nelem; j++) {
              if (scanf("%*[ ,]%lf", pixcrd+j) != 1) {
                ok = 0;
                break;
              }
            }
          }

          while (fgetc(stdin) != '\n');

          if (!ok) {
            wcsprintf("Input error, please try again.\n");
            continue;
          }

          wcsprintf("Pixel: ");
          for (int j = 0; j < nelem; j++) {
            wcsprintf("%s%14.9g", j?", ":"", pixcrd[j]);
          }
          wcsprintf("\n");

          double phi, theta;
          if ((status = wcsp2s(wcs+i, 1, nelem, pixcrd, imgcrd, &phi, &theta,
                               world, stat))) {
            wcsperr(wcs+i, "");

          } else {
            wcsprintf("Image: ");
            for (int j = 0; j < nelem; j++) {
              if (j == wcs[i].lng || j == wcs[i].lat) {
                // Print angles in fixed format.
                wcsprintf("%s%14.6f", j?", ":"", imgcrd[j]);
              } else {
                wcsprintf("%s%14.9g", j?", ":"", imgcrd[j]);
              }
            }

            wcsprintf("\nWorld: ");
            for (int j = 0; j < nelem; j++) {
              if (j == wcs[i].lng || j == wcs[i].lat) {
                // Print angles in fixed format.
                wcsprintf("%s%14.6f", j?", ":"", world[j]);
              } else {
                wcsprintf("%s%14.9g", j?", ":"", world[j]);
              }
            }
            wcsprintf("\n");
          }
        }
      }


      if (dowrld) {
        // Transform world coordinates.
        while (1) {
          wcsprintf("\nEnter %d world coordinate element%s: ", nelem,
            (nelem==1)?"":"s");
          int c = fgetc(stdin);
          if (c == EOF || c == '\n') {
            if (c == EOF) wcsprintf("\n");
            break;
          }
          ungetc(c, stdin);

          int ok;
          if ((ok = (scanf("%lf", world) == 1))) {
            for (int j = 1; j < nelem; j++) {
              if (scanf("%*[ ,]%lf", world+j) != 1) {
                ok = 0;
                break;
              }
            }
          }

          while (fgetc(stdin) != '\n');

          if (!ok) {
            wcsprintf("Input error, please try again.\n");
            continue;
          }

          wcsprintf("World: ");
          for (int j = 0; j < nelem; j++) {
            if (j == wcs[i].lng || j == wcs[i].lat) {
              // Print angles in fixed format.
              wcsprintf("%s%14.6f", j?", ":"", world[j]);
            } else {
              wcsprintf("%s%14.9g", j?", ":"", world[j]);
            }
          }
          wcsprintf("\n");

          double phi, theta;
          if ((status = wcss2p(wcs+i, 1, nelem, world, &phi, &theta, imgcrd,
                               pixcrd, stat))) {
            wcsperr(wcs+i, "");

          } else {
            wcsprintf("Image: ");
            for (int j = 0; j < nelem; j++) {
              if (j == wcs[i].lng || j == wcs[i].lat) {
                // Print angles in fixed format.
                wcsprintf("%s%14.6f", j?", ":"", imgcrd[j]);
              } else {
                wcsprintf("%s%14.9g", j?", ":"", imgcrd[j]);
              }
            }

            wcsprintf("\nPixel: ");
            for (int j = 0; j < nelem; j++) {
              wcsprintf("%s%14.9g", j?", ":"", pixcrd[j]);
            }
            wcsprintf("\n");
          }
        }
      }

      free(world);
      free(imgcrd);
      free(pixcrd);
      free(stat);
    }
  }

  fits_close_file(fptr, &fitstat);

  // Defeat spurious reporting of memory leaks.
  wcsvfree(&nwcs, &wcs);

  return 0;

cleanup:
  if (fitstat) fits_report_error(stderr, fitstat);
  fits_close_file(fptr, &fitstat);
  return 1;
}
